src/core/ext/transport/cronet/transport/cronet_transport.c

```diff
@@ -0,0 +1,583 @@
+  gpr_slice_buffer_add(&s->read_slicebuffer, read_data_slice);
+  grpc_slice_buffer_stream_init(&s->sbs, &s->read_slicebuffer, 0);
+  *s->recv_message = (grpc_byte_buffer *)&s->sbs;
+}
+
+int parse_grpc_header(const uint8_t *data) {
+  const uint8_t *p = data + 1;
+  uint32_t length = 0;
+  length |= ((uint8_t)*p++) << 24;
+  length |= ((uint8_t)*p++) << 16;
+  length |= ((uint8_t)*p++) << 8;
+  length |= ((uint8_t)*p++);
+  return length;
+}
+
+void on_read_completed(cronet_bidirectional_stream *stream, char *data,
+                       int count) {
+  stream_obj *s = (stream_obj *)stream->annotation;
+  GRPC_CRONET_TRACE(GPR_DEBUG,
+                    "R: on_read_completed count=%d, total=%d, remaining=%d",
+                    count, s->total_read_bytes, s->remaining_read_bytes);
+  if (count > 0) {
+    GPR_ASSERT(s->recv_message);
+    s->remaining_read_bytes -= count;
+    next_recv_step(s, ON_READ_COMPLETE);
+  } else {
+    s->read_closed = true;
+    next_recv_step(s, ON_READ_COMPLETE);
+  }
+}
+
+void on_response_headers_received(
+    cronet_bidirectional_stream *stream,
+    const cronet_bidirectional_stream_header_array *headers,
+    const char *negotiated_protocol) {
+  GRPC_CRONET_TRACE(GPR_DEBUG, "R: on_response_headers_received");
+  stream_obj *s = (stream_obj *)stream->annotation;
+  enqueue_callbacks(s->callback_list[CB_RECV_INITIAL_METADATA]);
+  s->response_headers_received = true;
+  next_recv_step(s, ON_RESPONSE_HEADERS_RECEIVED);
+}
+
+void on_request_headers_sent(cronet_bidirectional_stream *stream) {
+  GRPC_CRONET_TRACE(GPR_DEBUG, "W: on_request_headers_sent");
+  stream_obj *s = (stream_obj *)stream->annotation;
+  enqueue_callbacks(s->callback_list[CB_SEND_INITIAL_METADATA]);
+  cronet_send_state = CRONET_SEND_HEADER;
+  next_send_step(s);
+}
+
+// Callback function pointers (invoked by cronet in response to events)
+cronet_bidirectional_stream_callback callbacks = {on_request_headers_sent,
+                                                  on_response_headers_received,
+                                                  on_read_completed,
+                                                  on_write_completed,
+                                                  on_response_trailers_received,
+                                                  on_succeded,
+                                                  on_failed,
+                                                  on_canceled};
+
+void invoke_closing_callback(stream_obj *s) {
+  grpc_chttp2_incoming_metadata_buffer_publish(&s->imb,
+                                               s->recv_trailing_metadata);
+  if (s->callback_list[CB_RECV_TRAILING_METADATA]) {
+    enqueue_callbacks(s->callback_list[CB_RECV_TRAILING_METADATA]);
+  }
+}
+
+// This is invoked from perform_stream_op, and all on_xxxx callbacks.
+void next_recv_step(stream_obj *s, enum e_caller caller) {
+  gpr_mu_lock(&s->recv_mu);
+  switch (cronet_recv_state) {
```

Comment:
This variable being global looks like a massive bug: what happens when we have two streams live on a TCP connection?